Use a Python entry point for 'display_urdf'

cottsay opened the pull request:
On many platforms, the unversioned python executable is not Python 2. A better solution here would be to change this package to use an entry point, as shebang python scripts don't work correctly on Windows.

This same change applies to Noetic, but there isn't a Noetic-specific branch to make the change against.

sloretz reviewed (dismissed):
> A better solution here would be to change this package to use an entry point, as shebang python scripts don't work correctly on Windows.

Agreed. I'd much rather see that fix get implemented: remove the shebang, make the script not executable, and add a `console_scripts` entrypoint in the `setup.py` for ROS 2.


> This same change applies to Noetic, but there isn't a Noetic-specific branch to make the change against.

I don't think any changes are needed in ROS 1. ROS Melodic and Noetic [use `catkin_install_python()` to rewrite the shebang](http://wiki.ros.org/UsingPython3/SourceCodeChanges#Changing_shebangs) in the devel space and when it is installed.

https://github.com/ros/urdf_parser_py/blob/b6a9726c976933fb8cec0b5b872101984ec131dd/CMakeLists.txt#L9-L12

cottsay committed:
Use a Python entry point for 'display_urdf' (#64)

This will make the script work on Windows, and will ensure that the
correct Python interpreter is always used.

Signed-off-by: Scott K Logan <logans@cottsay.net>

cottsay committed:
Make 'file' argument to display_urdf required (#64)

The underlying operation, fetching the URDF from the parameter server,
is no longer supported in ROS 2.

Signed-off-by: Scott K Logan <logans@cottsay.net>

cottsay committed:
Move the 'display_urdf' script into the Python module (#64)

The change following this one will modify the script to use a Python
entry point.

Signed-off-by: Scott K Logan <logans@cottsay.net>

cottsay commented:
>  remove the shebang, make the script not executable, and add a `console_scripts` entrypoint in the `setup.py`

Done. I split it into three commits. The first resolves a ROS 2 bug, the second does the move, and the third fixes up the script as an entry point. We should rebase-merge this to preserve the change history of `display_urdf`.

> I don't think any changes are needed in ROS 1

You're right, I forgot about that. Good call - no changes necessary for ROS 1.

clalancette commented on src/urdf_parser_py/display_urdf.py:
```diff
@@ -1,20 +1,20 @@
-args = parser.parse_args()
 
-robot = URDF.from_xml_string(args.file.read())
+def main(argv=None):
```
Nit: in ROS 2 we typically use `args` instead of `argv` (though we aren't 100% consistent, sigh).

sloretz replied:
While `args/argv` argument has become widespread, I'm pretty sure it doesn't actually work. I think scripts still have to get arguments from `sys.argv`.

The template for `console_scripts` doesn't pass anything - see the function call on the entry point is just`()`.
https://github.com/pypa/setuptools/blob/990237c64404876791dc5985e7cfd260852dac0e/setuptools/command/easy_install.py#L2069

It's also not mentioned in the docs.
https://python-packaging.readthedocs.io/en/latest/command-line-scripts.html#the-console-scripts-entry-point

clalancette reviewed (approved):
This looks good to me; I'll leave it to you whether you want to fix the nit or not.

I'd prefer to wait to merge this until @sloretz has had another chance to look it over.

sloretz commented on src/urdf_parser_py/display_urdf.py:
```diff
@@ -0,0 +1,20 @@
+
+
+def main(argv=None):
+    parser = argparse.ArgumentParser(argv, usage='Load an URDF file')
```
I'd be surprised if argv is ever not `None`. This happens to be ok because it looks like the first argument is `prog` which is supposed to be the name of the program, else `sys.argv[0]` if `None`. https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser

Still I think this should be removed for clarity.

clalancette commented:
@cottsay Friendly ping here; I think this is close to getting done, with just a couple more nits to fix from @sloretz .

cottsay commented:
Okay, I dropped the `argv` stuff. Again, please rebase-merge this PR when you're happy with it - I updated the commits to include a reference to this PR so they look good when rebase-merged.

clalancette commented:
I'm going to go ahead and merge this, thanks for the patch @cottsay
